core/src/executor/datafusion/datafusion_processor.rs

```diff
@@ -116,23 +119,37 @@ impl DatafusionProcessor {
 
     pub async fn execute(&mut self) -> Result<(Vec<SendableRecordBatchStream>, Schema)> {
         self.register_tables()?;
+        let schema = self.datafusion_task_ctx.input_schema.take().unwrap();
         let df = self.ctx.sql(&self.datafusion_task_ctx.exec_sql).await?;
-        let physical_plan = df.create_physical_plan().await?;
-        let batchs =
-            if physical_plan.output_partitioning().partition_count() != self.target_partitions {
-                let physical_plan: Arc<dyn ExecutionPlan + 'static> =
-                    Arc::new(RepartitionExec::try_new(
-                        physical_plan,
-                        Partitioning::RoundRobinBatch(self.target_partitions),
-                    )?);
-                execute_stream_partitioned(physical_plan, self.ctx.task_ctx())?
-            } else {
-                execute_stream_partitioned(physical_plan, self.ctx.task_ctx())?
-            };
-        Ok((
-            batchs,
-            self.datafusion_task_ctx.input_schema.take().unwrap(),
-        ))
+        let mut physical_plan = df.create_physical_plan().await?;
+        if self.partition_spec.is_unpartitioned()
+            && physical_plan.output_partitioning().partition_count() != self.target_partitions
+        {
+            let partitioning = Partitioning::RoundRobinBatch(self.target_partitions);
+            physical_plan = Arc::new(RepartitionExec::try_new(physical_plan, partitioning)?);
+        } else if !self.partition_spec.is_unpartitioned() {
+            let partitioning = Partitioning::Hash(
+                self.partition_spec
```

Comment:
I think we can modify this logic based on https://github.com/risingwavelabs/risingwave/blob/776d84746bd11c92b6443b4fc96b5db18f49d9f8/src/frontend/src/handler/create_sink.rs#L417-L435 to avoid unnecessary hash shuffle.